rfcs/20240722-group-query-attention/README.md

```diff
@@ -0,0 +1,215 @@
+
+Cons:
+
+1. Need to add more operations to oneDNN graph operation set.
+2. The implementation in the community may change. Even in the Huggingface code,
+   another alternative implementation is explicitly mentioned (via
+   `torch.repeat_interleave`). Once the implementation changes, the above pros
+   will become invalid.
+
+### Option 2 (recommended)
+
+Add new patterns. We can reshape Query, Key and Values tensors from 4D to 5D and
+leverage the broadcasting semantics of MatMul operation to perform the dot-products.
+  
+1. Reshape Query from 4D shape (N, H, S, D) to 5D shape (N, G, H / G, S, D).
+2. Reshape Key from 4D shape (H, G, S, D) to 5D shape (N, G, 1, S, D).
+3. Performance 5D matmul between (N, G, H / G, S, D) and (N, G, 1, S, D). The
+   third dimension will be broadcasted from 1 to `H / G` automatically per the
+   broadcasting rule of MatMul operation.
+4. Similar reshape and broadcasting can also be applied to the dot-product of
+   Value.
+
+Here is the diagram:
+![option2](option2.png)
+
+Pros:
+
+1. No change to API.
+2. The pattern looks simpler than the pattern in option 1.
+
+Cons:
+
+1. The pattern is less intuitive from GQA definition.
+2. The pattern cannot be used to optimize a framework graph directly. Frameworks
+   will have to implement GQA fusion by themselves and leverage this option to
+   optimized the fused GQA.
+
+### Option 3
+
+We can extend the MatMul broadcasting rules to support group broadcast.
+Currently, MatMul operation supports below broadcasting rules.
+
+For single 2D matrix multiplication:
+
+| Matrix A | Matrix B | Matrix C = A x B |
+| --       | --       | --               |
+| M x K    | K x N    | M x N            |
+
+For batched matrix multiplications:
+
+| Matrix A   | Matrix B   | Matrix C = A x B |
+| --         | --         | --               |
+| B x M x K  | B x K x N  | B x M x N        |
+| B x M x K  | 1 x K x N  | B x M x N        |
+| 1 x M x K  | B x K x N  | B x M x N        |
+
+This can be extended to multiple batch dimensions matrix multiplications:
+
+| Matrix A              | Matrix B              | Matrix C = A x B      |
+| --                    | --                    | --                    |
+| B1 x 1 x B3 x M x K   | B1 x B2 x 1 x M x K   | B1 x B2 x B3 x M x N  |
+
+This RFC proposes to further extend it to support group broadcast:
+
+| Matrix A              | Matrix B                | Matrix C = A x B      |
+| --                    | --                      | --                    |
+| B1 x 1 x B3 x M x K   | 1 x B2 x (B3/c) x K x N | B1 x B2 x B3 x M x N  |
+
+where, c is a factor of B3.
+
+This rule looks uncommon and is not supported by the typical broadcasting rules
+(see broadcasting in
+[ONNX](https://github.com/onnx/onnx/blob/main/docs/Broadcasting.md) and
+[NumPy](https://numpy.org/doc/stable/user/basics.broadcasting.html#general-broadcasting-rules).),
+but actually it's added to the MatMul operation of cuDNN in order to support
```

Comment:
I'd like to add the link here for reference: https://docs.nvidia.com/deeplearning/cudnn/latest/api/cudnn-graph-library.html#cudnn-backend-operation-matmul-descriptor .

Reply:
Added, thanks.